### PR TITLE
Expose node replacement mapping functionality in the API

### PR DESCRIPTION
**Pull request by Treer**
From [discussion in #2](https://github.com/bell07/minetest-schemlib/pull/2#issuecomment-468895693)

### CODE REVIEW

File: plan.lua [hunk 1]
```diff
@@ -339,6 +339,43 @@ function plan_class:apply_flood_with_air(add_max, add_min, add_top)
...
+		new_mapping = { 
+			-- required fields:
+			name     = to,
+			node_def = minetest.registered_nodes[to],
+			-- optional fields:
+			param2          = param2,
+			meta            = meta,
+			prob            = prob,
+			custom_function = custom_function
+		}
```
Comment (bell07) [inline plan.lua]: `cost_item` is missed!

File: plan.lua [hunk 2]
```diff
@@ -339,6 +339,43 @@ function plan_class:apply_flood_with_air(add_max, add_min, add_top)
 	dprint("flatting plan done")
 end
 
+-----------------------------------------------------
+-- Get an existing node-replacement mapping, or nil 
+-----------------------------------------------------
+-- Returns a map entry table which contains name, node_def, and optional extra fields.
+-- Returns nil if the plan had no mapping for the given node_name
+function plan_class:get_replacement(node_name)
+	return self.mapping_cache[node_name]
```
[the remark this answers is not in the document]
Comment (Treer) [reply plan.lua]: Surely only set_replacement() should create mappings, get_replacement is a read operation.

File: README.md [hunk 3]
```diff
@@ -47,7 +49,10 @@ In this stage the world interaction methods like place node are possible.
   - plan_obj:get_node(plan_pos)             - get a node from plan
   - plan_obj:del_node(plan_pos)             - delete a node from plan
   - plan_obj:get_random_plan_pos()          - get a random existing plan_pos from plan
-  - plan_obj:read_from_schem_file(file)     - read from WorldEdit or mts file
+  - plan_obj:read_from_schem_file(file[,replacments])    - read from WorldEdit or mts file
+    - replacments = `{["old_name"] = "convert_to", ...}` - optional, and only used by mts files.
+  - plan_obj:get_replacement(node_name)     - get a map entry table specifying the name, node_def, and any optional extra fields that nodes of type node_name will be replaced with. Returns nil if the plan_obj has no replacement mapping for node_name.
+  - plan_obj:set_replacement(from_node_name, to[,param2][,meta][,prob][,custom_function]) - add a node replacement mapping. Only the 'from_node_name' and 'to' arguments are required. 'to' can be either a node name, or a map entry table.
```
[the remark this answers is not in the document]
Comment (Treer) [reply README.md]: ahh, so you want to get rid of set_replacement(), and just adjust tables returned by get_replacement.

File: plan.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (bell07) [reply plan.lua]: The get_replacement() is basically a wrapper for mapping.map() and does use the mapping_cache

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bell07) [reply README.md]: set_replacement() is still useful for creation the fully custom mappings without getting the default values trough get_replacements() / mapping.map()

File: plan.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (bell07) [reply plan.lua]: Maybe an other name is beter? get_mapping/set_mapping? What do you think?

**Comment by bell07** [comment]
the topic did not let me go today. Working on cache table is wrong. The replacements needs to be changeable before the cache is filled. Working on cache we get issues if rotation is used after mapping is set (at least with doors) and may other issues.
Now I have the idea to add new table with custom mapping entries like hard-coded in mapping lua.

Let me do it. I continue work on this PR tonight

**Comment by bell07** [comment]
I did it. Pls. look to my master for changes.

I removed the plan:get_replacement() because plan.data.replacements can be accessed directly.
The set_replacement is not really necessary for the same reason, but I keep it, but with changed interface.

`plan:set_replacement('default:stonebrick', {name = 'default:stone_with_gold'})` does work as expected.

by the way, I moved the cost_item determination out from mapping handling to new function mapping.get_cost_item() because useful only for "Builder-NPC" mods. stujones11/minetest-npcf updated.

Just for information, for example you like to fill chests with content: Set "meta" for chest in mapping leads to the same content in all chests, but set custom_function can be used to fill chests with different items.

**Pull request merged**